arc_solvers/nn/util.py

```diff
@@ -32,3 +35,130 @@ def masked_mean(tensor, dim, mask):
     return mean_tensor
 
 
+
+def seq2vec_seq_aggregate(seq_tensor, mask, aggregate, bidirectional, dim=1):
+    """
+        Takes the aggregation of sequence tensor
+
+        :param seq_tensor: Batched sequence requires [batch, seq, hs]
+        :param mask: binary mask with shape batch, seq_len, 1
+        :param aggregate: max, avg, sum
+        :param dim: The dimension to take the max. for batch, seq, hs it is 1
+        :return:
+    """
+
+    seq_tensor_masked = seq_tensor * mask.unsqueeze(-1)
+    aggr_func = None
+    if aggregate == "last":
+        raise NotImplemented("This is currently not supported.")
+        seq = allennlp.nn.util.get_final_encoder_states(seq_tensor, mask, bidirectional)
+    elif aggregate == "max":
+        aggr_func = torch.max
+        seq, _ = aggr_func(seq_tensor_masked, dim=dim)
+    elif aggregate == "min":
+        aggr_func = torch.min
+        seq, _ = aggr_func(seq_tensor_masked, dim=dim)
+    elif aggregate == "sum":
+        aggr_func = torch.sum
+        seq = aggr_func(seq_tensor_masked, dim=dim)
+    elif aggregate == "avg":
+        aggr_func = torch.sum
+        seq = aggr_func(seq_tensor_masked, dim=dim)
+        seq_lens = torch.sum(mask, dim=dim)  # this returns batch_size, 1
+        seq = seq / seq_lens.view([-1, 1])
+
+    return seq
+
+
+def embedd_encode_and_aggregate_text_field(question: Dict[str, torch.LongTensor],
+                                           text_field_embedder,
+                                           embeddings_dropout,
+                                           encoder,
+                                           aggregation_type,
+                                           get_last_states=False):
+    """
+    Given a batched token ids (2D) runs embeddings lookup with dropout, context encoding and aggregation
+    :param question:
+    :param text_field_embedder: The embedder to be used for embedding lookup
+    :param embeddings_dropout: Dropout
+    :param encoder: Context encoder
+    :param aggregation_type: The type of aggregation - max, sum, avg, last
+    :param get_last_states: If it should return the last states.
+    :return:
+    """
+    embedded_question = text_field_embedder(question)
+    question_mask = get_text_field_mask(question).float()
+    embedded_question = embeddings_dropout(embedded_question)
+
+    encoded_question = encoder(embedded_question, question_mask)
+
+    # aggregate sequences to a single item
+    encoded_question_aggregated = seq2vec_seq_aggregate(encoded_question, question_mask, aggregation_type,
+                                                        None, 1)  # bs X d
+
+    last_hidden_states = None
+    if get_last_states:
+        last_hidden_states = allennlp.nn.util.get_final_encoder_states(encoded_question, question_mask, encoder.is_bidirectional())
+
+    return encoded_question_aggregated, last_hidden_states
```

Comment:
Why does it also return the last hidden state? It seems like that makes a lot of assumptions about what `encoder()` does. What if `encoder()` is a convnet?

Reply:
Looks like the only place where this is used, the last hidden states are ignored anyways?

Reply:
The encoder, in this case, is considered to be RNN based. But the hidden states are not used in this project, so I will remove them.